File: scalechain-crypto/src/main/scala/io/scalechain/crypto/HashFunctions.scala
package io.scalechain.crypto

import java.security.MessageDigest
import org.spongycastle.crypto.digests.RIPEMD160Digest

trait HashValue {
  val value: Array[Byte]
}

case class SHA1(bytes:Array[Byte]) extends HashValue {
  override val value = bytes
}

case class SHA256(bytes:Array[Byte]) extends HashValue {
  override val value = bytes
}

case class RIPEMD160(bytes:Array[Byte]) extends HashValue {
  override val value = bytes
}

case class Hash160(bytes:Array[Byte]) extends HashValue {
  override val value = bytes
}

case class Hash256(bytes:Array[Byte]) extends HashValue {
  override val value = bytes
}


/**
 * Created by kangmo on 11/11/15.
 */
object HashFunctions {
  /**
   *
   * @param input
   * @return
   */
  def sha1(input: Array[Byte]) : SHA1 = {
    val md = MessageDigest.getInstance("SHA-1")
    SHA1( md.digest(input) )
  }

  /**
   *
   * @param input
   * @return
   */
  def sha256(input: Array[Byte]) : SHA256 = {
    val md = MessageDigest.getInstance("SHA-256")
    SHA256( md.digest(input) )
  }

  /**
   *
   * @param input
   * @return
   */
  def ripemd160(input: Array[Byte]) : RIPEMD160 = {
    val md = new RIPEMD160Digest()
    md.update(input, 0, input.length)
    val out = Array.fill[Byte](md.getDigestSize())(0)
    md.doFinal(out, 0)
    RIPEMD160(out)
  }

  /** Return RIPEMD160(SHA256(x)) hash
   *
   * @param input
   * @return
   */
  def hash160(input: Array[Byte]) : Hash160 = {
    Hash160( ripemd160( sha256(input).value ).value )
  }

  /** Return SHA256(SHA256(x)) hash
   *
   * @param input
   * @return
   */
  def hash256(input: Array[Byte]) : Hash256 = {
    Hash256( sha256( sha256(input).value ).value )
  }
}

File: scalechain-crypto/src/main/kotlin/io/scalechain/crypto/HashFunctions.kt
package io.scalechain.crypto

import io.netty.buffer.ByteBuf
import io.scalechain.util.Bytes
import java.security.MessageDigest
import org.spongycastle.crypto.digests.RIPEMD160Digest

interface HashValue {
  val value: Bytes
}

data class SHA1(override val value:Bytes) : HashValue

data class SHA256(override val value:Bytes) : HashValue

data class RIPEMD160(override val value:Bytes) : HashValue

data class Hash160(override val value:Bytes) : HashValue

data class Hash256(override val value:Bytes) : HashValue

/**
 * Created by kangmo on 11/11/15.
 */
object HashFunctions {
  /**
   *
   * @param input
   * @return
   */
  fun sha1(input: ByteArray) : SHA1 {
    val sha1md = MessageDigest.getInstance("SHA-1")
    return SHA1( Bytes( sha1md.digest(input) ) )
  }

  /**
   *
   * @param input
   * @return
   */
  fun sha256(input: ByteArray) : SHA256 {
    val sha256md = MessageDigest.getInstance("SHA-256")
    return SHA256( Bytes( sha256md.digest(input) ) )
  }

  fun sha256(input: ByteArray, offset : Int, length : Int) : SHA256 {
    val sha256md = MessageDigest.getInstance("SHA-256")
    sha256md.update(input, offset, length)
    return SHA256( Bytes( sha256md.digest() ) )
  }


  /**
   *
   * @param input
   * @return
   */
  fun ripemd160(input: ByteArray) : RIPEMD160 {
    val md = RIPEMD160Digest()
    md.update(input, 0, input.size)
    val out = ByteArray(md.getDigestSize(), {0})
    md.doFinal(out, 0)
    return RIPEMD160( Bytes(out) )
  }

  /** Return RIPEMD160(SHA256(x)) hash
   *
   * @param input
   * @return
   */
  fun hash160(input: ByteArray) : Hash160 {
    return Hash160( ripemd160(sha256(input).value.array ).value )
  }

  /** Return SHA256(SHA256(x)) hash
   *
   * @param input
   * @return
   */
  fun hash256(input: ByteArray) : Hash256 {
    return Hash256( sha256( sha256(input).value.array ).value )
  }

  fun hash256(input: ByteArray, offset : Int, length : Int) : Hash256 {
    return Hash256( sha256( sha256(input, offset, length).value.array ).value )
  }

}
